### PR TITLE
fix(Scripts/Spells): Flag of Ownership

### PR DESCRIPTION
**Pull request by xaenerys0**
…lization

## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/18659

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

This is how I tested the changes:
1. Login with an alliance member in one client
2. Login with a horde member in another client
3. Teleport them to the same spot where PvP is enabled (make sure PvP is enabled for both characters), I used Outland
4. Give the flag of ownership to the killer, then kill the member of the other faction
5. Cast the Flag of Ownership and observe the text

image

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
7.
8.

## Known Issues and TODO List:

- [ ] Potentially adjust how I am calling the placeholder variables. I could not get it working without the two %s placeholders. I am new to C++ so I could be missing something obvious, but a possible refactor.
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by fangshun2004** [comment]
The data in the broadcast_text and broadcast_text_locale is sniffed and cannot be modified under normal circumstances

**Comment by xaenerys0** [comment]
> The data in the broadcast_text and broadcast_text_locale is sniffed and cannot be modified under normal circumstances

So the issue I am facing now is that using the broadcast_text unmodified duplicates the player/caster name: %s plants the Flag of Ownership in the corpse of $n.

%s is the player name, and outside of the broadcast text it is also printed, so the result is: PlayerName PlayerName plants the Flag of Ownership in the corpse of TargetName. This is what initially led me to my original solution of modifying that text before I knew about the limitations of updating that table.

File: src/server/scripts/Spells/spell_generic.cpp [hunk 1]
```diff
@@ -48,6 +48,11 @@
 //  see: https://github.com/azerothcore/azerothcore-wotlk/issues/9766
 #include "GridNotifiersImpl.h"
 
+enum broadcastTextIDs
+{
+    BROADCAST_TEXT_FLAG_OF_OWNERSHIP     = 28008, // The Flag of Ownership
```
Comment (heyitsbench) [inline src/server/scripts/Spells/spell_generic.cpp:53]: This comment is redundant (but appreciated).

Comment (xaenerys0) [reply src/server/scripts/Spells/spell_generic.cpp:53]: Would you like it removed? I'm unsure due to the appreciation. I can do so once I know a solution to the TextEmote issue. Thanks!

Comment (heyitsbench) [reply src/server/scripts/Spells/spell_generic.cpp:53]: It should be removed IMO, yes.

**Comment by xaenerys0** [comment]
> > So the issue I am facing now is that using the broadcast_text unmodified duplicates the player/caster name: %s plants the Flag of Ownership in the corpse of $n.
> 
> Perhaps something other than `TextEmote` can be used? Maybe someone more familiar can chime in and throw a suggestion to resolve the issue.

I have a working solution local that does the same thing I did with the $n placeholder. However, I am not sure if this is the best approach. I can commit and have this ready for review unless someone else has any suggestions for alternatives to `TextEmote`.

File: src/server/scripts/Spells/spell_generic.cpp [hunk 2]
```diff
@@ -135,8 +140,28 @@ class spell_the_flag_of_ownership : public SpellScript
...
+
+        // TextEmote without the string replacement duplicates the character name
+        pos = formatString.find("%s ");
+        if (pos != std::string::npos) {
+            formatString.replace(pos, 3, "");
+        }
+
         char buff[100];
-        snprintf(buff, sizeof(buff), "%s plants the Flag of Ownership in the corpse of %s.", caster->GetName().c_str(), target->GetName().c_str());
+        snprintf(buff, sizeof(buff), formatString.c_str());
```
Comment (Kitzunu) [inline src/server/scripts/Spells/spell_generic.cpp:164]: You can replace it all to

```cpp
        LocaleConstant loc_idx = caster->ToPlayer()->GetSession()->GetSessionDbLocaleIndex();
        BroadcastText const* bct = sObjectMgr->GetBroadcastText(BROADCAST_TEXT_FLAG_OF_OWNERSHIP);
        std::string bctMsg = Acore::StringFormat(bct->GetText(loc_idx, caster->getGender()), caster->GetName().c_str(), target->GetName().c_str());
        caster->Talk(bctMsg, CHAT_MSG_MONSTER_EMOTE, LANG_UNIVERSAL, sWorld->getFloatConfig(CONFIG_LISTEN_RANGE_TEXTEMOTE), target);
```

image

**Comment by Kitzunu** [comment]
@xaenerys0 Status

**Comment by xaenerys0** [comment]
@Kitzunu will do it today (has some life stuff going on since last comments).

File: src/server/scripts/Spells/spell_generic.cpp [hunk 3]
```diff
@@ -114,6 +114,12 @@ class spell_gen_model_visible : public AuraScript
     bool _hasFlag;
 };
 
+enum flagOfOwnership
```
Comment (Kitzunu) [inline src/server/scripts/Spells/spell_generic.cpp:117]:
```suggestion
enum FlagOfOwnership
```

**Review by Kitzunu — APPROVED** [review]

**Comment by Kitzunu** [comment]
Nice. A tip for next time is to create a new branch when making PRs and not modify your `master` branch on your fork. As it can mess things up for you in the future. So now I suggest you to nuke or clean your `master` branch to be 1-1 with the main repo :D

**Pull request merged**